### PR TITLE
Convert all regression tests to use ESMF managed threading

### LINKED ISSUE #1111: Using ESMF managed threading in UFS
This EPIC is to track the development of using the ESMF managed threading in UFS.

1) Update to ESMF 8.2.0 release.
2) Adding the code updates to enable testing the ESMF managed threading in UFS.
3) Update the intel/impi to Intel 2021 to resolve the mpi_alltoall scaling issue.
4) Enable the ESMF managed threading in UFS, add regression tests to use the capability

Documentation from Gerhard is available [here](https://docs.google.com/presentation/d/1-LraZX9GryW5t0z5waIEYVofyR0qi-4uJCajSNKmSVM).

### PR DESCRIPTION
**Pull request by DusanJovic-NOAA**
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR
are specified below.

- [ ] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.

- [ ] New or updated input data is required by this PR. If checked, please work with the code managers to update input data sets on all platforms.

## Instructions: All subsequent sections of text should be filled in as appropriate.

The information provided below allows the code managers to understand the changes relevant to this PR, whether those changes are in the ufs-weather-model repository or in a subcomponent repository. Ufs-weather-model code managers will use the information provided to add any applicable labels, assign reviewers and place it in the Commit Queue. Once the PR is in the Commit Queue, it is the PR owner's responsibility to keep the PR up-to-date with the develop branch of ufs-weather-model.

## Description

This PR updates all regression tests to use ESMF managed threading. Summary of changes:
- in default_vars.sh set  the number of threads to be used for all components (atm_omp_num_threads, ocn_omp_num_threads, ...) to 1  (no threading by default)
- some of these values are overwritten in 'threading' config files to specify the number of threads being used for some of the components
- these values are then used in the 'compute_petbounds_and_tasks' function in rt_utils.sh to compute required PEs used by each component
- all nems.configure.*.IN templates are updated to specify the values of atm_omp_num_threads, ... and set globalResourceControl:  to true to enable use of ESMF managed threading
- setting of variable TASKS is removed from all tests and default_vars.sh , it is now computed.
- minor cleanups.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues must always be created before starting work on a PR branch!)
- fixes #1111

## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel
- [x] cheyenne.gnu
- [x] gaea.intel
- [x] jet.intel
- [x] wcoss2.intel
- [x] acorn.intel
- [ ] opnReqTest for newly added/changed feature
- [ ] CI

## Dependencies

Do PRs in upstream repositories need to be merged first? No

### CODE REVIEW

File: tests/parm/nems.configure.hafs_atm_wav.IN [hunk 1]
```diff
@@ -41,36 +44,10 @@ ATM_attributes::
   write_restart_at_endofrun = .true.
 ::
 
-# OCN #
```
Comment (DeniseWorthen) [inline tests/parm/nems.configure.hafs_atm_wav.IN:44]: What happened to these settings?

Comment (DusanJovic-NOAA) [reply tests/parm/nems.configure.hafs_atm_wav.IN:44]: This test does not use OCN component.

Comment (DeniseWorthen) [reply tests/parm/nems.configure.hafs_atm_wav.IN:44]: Ah---good catch!

Comment (DusanJovic-NOAA) [reply tests/parm/nems.configure.hafs_atm_wav.IN:44]: Now that I look more carefully, two arrays at the end of this file `ocn_export_fields` and `ocn_import_fields` are probably not needed, but I'm not sure.
@BinLiu-NOAA

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA We are working on #1488. I think we should be able to merge in #1488 tomorrow. Can I check if you can combine in #1541 and #1522 to this pr ? all these PRs are no baseline changes.

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA We are working on #1488. I think we should be able to merge in #1488 tomorrow. Can I check if you can combine in #1541 and #1522 to this pr ? all these PRs are no baseline changes.

I think we can. Once #1488 is merged please ask #1541 and #1522 PR owners to update their branches and I'll merge them into this PR.

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA more time is needed on #1488. We like to start working on this pr. @binli2337 can you work to combine in #1541 to this pr?

**Comment by jkbk2004** [comment]
@dustinswales if you are working today, can we talk about how to combine in #1522 to this pr?

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA I was in wrong pr. We can start working on this pr alone. Sounds like Bin and Dustin are not available. Bin's PR is https://github.com/ufs-community/ufs-weather-model/pull/1541. Dustins' is https://github.com/ufs-community/ufs-weather-model/pull/1522. https://github.com/ufs-community/ufs-weather-model/pull/1541 is simple. I think you can force to push the https://github.com/ufs-community/ufs-weather-model/pull/1541 change into this pr. But feel free.

**Comment by DusanJovic-NOAA** [comment]
I just merged #1541 into this PR.

**Comment by jkbk2004** [comment]
> I just merged #1541 into this PR.

Cool! I will start testing from hera.

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA We used to see this issue id_file issue on aux.yml. I am bothered now. Can you comment out line 20 to 37 in .github/workflows/aux.yml ? Those were used to handle previous run-ci label that we don't need anymore. Or I can directly exercise with aux.yml.

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA We used to see this issue id_file issue on aux.yml. I am bothered now. Can you comment out line 20 to 37 in .github/workflows/aux.yml ? Those were used to handle previous run-ci label that we don't need anymore. Or I can directly exercise with aux.yml.

See https://github.com/ufs-community/ufs-weather-model/pull/1523/commits/ed7fd950234dcbfbd9e8383e1ea6eb52295c6388

**Comment by BrianCurtis-NOAA** [comment]
***Automated RT Failure Notification***
Machine: gaea
Compiler: intel
Job: RT
[RT] Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/1151609766/20221222203008/ufs-weather-model
[RT] Error: Test control_c384 028 failed in run_test failed 
[RT] Error: Test control_c384gdas 029 failed in run_test failed 
Please make changes and add the following label back: gaea-intel-RT

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA Can you access to gaea? I think those two failed cases are hanging on gaea. I see same issue on cheyenne.

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA Can you access to gaea? I think those two failed cases are hanging on gaea. I see same issue on cheyenne.

Regression tests passed on Gaea after I changed c384 tests to use 1 thread.

Which tests are failing on Cheyenne?

**Comment by jkbk2004** [comment]
> > @DusanJovic-NOAA Can you access to gaea? I think those two failed cases are hanging on gaea. I see same issue on cheyenne.
> 
> Regression tests passed on Gaea after I changed c384 tests to use 1 thread.
> 
> Which tests are failing on Cheyenne?

cpld_restart_c192_p8, control_c384, and control_c384gdas

**Comment by DusanJovic-NOAA** [comment]
> > > @DusanJovic-NOAA Can you access to gaea? I think those two failed cases are hanging on gaea. I see same issue on cheyenne.
> > 
> > 
> > Regression tests passed on Gaea after I changed c384 tests to use 1 thread.
> > Which tests are failing on Cheyenne?
> 
> cpld_restart_c192_p8, control_c384, and control_c384gdas

For two c384 tests try to change the number of threads to 1:
```diff
diff --git a/tests/default_vars.sh b/tests/default_vars.sh
index d3ce19bf..e2e7242c 100755
--- a/tests/default_vars.sh
+++ b/tests/default_vars.sh
@@ -206,7 +206,7 @@ elif [[ $MACHINE_ID = cheyenne.* ]]; then
 
   INPES_dflt=3 ; JNPES_dflt=8
   INPES_thrd=3 ; JNPES_thrd=4
-  INPES_c384=8 ; JNPES_c384=6  ; THRD_c384=2
+  INPES_c384=8 ; JNPES_c384=6  ; THRD_c384=1
   INPES_c768=8 ; JNPES_c768=16 ; THRD_c768=2
 
   THRD_cpl_atmw_gdas=3
```
I'm not sure about  cpld_restart_c192_p8. did cpld_control_c192_p8 pass? If it did I do not see why would cpld_restart_c192_p8 fail.

**Comment by jkbk2004** [comment]
Ok! I will give a try.

**Comment by DeniseWorthen** [comment]
@jkbk2004 I was able to run both c192 tests on cheyenne w/o issue (using the original PR). What failure did you get on Cheynne?

**Comment by jkbk2004** [comment]
> @jkbk2004 I was able to run both c192 tests on cheyenne w/o issue (using the original PR). What failure did you get on Cheynne?

All others pass ok now but control_c384 still issue on cheyenne. let me try one time.

**Comment by jkbk2004** [comment]
@DeniseWorthen all others good but can you give a try to run control_c384 on cheyenne? still problem on my side

**Comment by DeniseWorthen** [comment]
@jkbk2004 I was able to run control_c384 but not control_c384gdas this time.

**Comment by DusanJovic-NOAA** [comment]
What kind of error do you see?

**Comment by jkbk2004** [comment]
I am trying to run with THRD_c384=1 on cheyenne now.

**Comment by DeniseWorthen** [comment]
> What kind of error do you see?

+ mpiexec_mpt -p %g: -np 648 omplace ./fv3.exe
MPT: shepherd terminated: r13i2n12.ib0.cheyenne.ucar.edu - job aborting

**Comment by jkbk2004** [comment]
I don't get control_c384gdas runs but control_c384 fails. @DeniseWorthen Can we agree on turning off control_c384 on cheyenne? If we agree, then I will create an issue to follow on later.

**Comment by DeniseWorthen** [comment]
@jkbk2004 yes, I think that would be OK.

**Comment by jkbk2004** [comment]
We suspect the cheyenne resource issue with control_c384. The issue #1549 was created to follow on. We can start merging process with final approvals.

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Review by DeniseWorthen — APPROVED** [review]

**Pull request merged**